### ISSUE TITLE
need for assertion in pick?

### ISSUE THREAD

**Issue by xrmx** [martinogden/django-banner-rotator#9]
Just got ~50 assertion in production, no fun. These are the actual banners on my site, l is the list of all the banner weight of the same place.

``` python
>>> l = [10, 10, 5, 3, 5, 10, 9]
>>> s = sum(l)
>>> s
52
>>> bias = [x/float(s) for x in l]
>>> bias
[0.19230769230769232, 0.19230769230769232, 0.09615384615384616, 0.057692307692307696, 0.09615384615384616, 0.19230769230769232, 0.17307692307692307]
>>> sum(bias)
1.0

>>> from random import random
>>> from decimal import Decimal
>>> number = Decimal("%.18f" % random())
>>> current = Decimal(0)
>>> i = 0
>>> for x in bias:
...   current += Decimal(x)
...   if number <= current:
...     print l[i]
...   i += 1
9
```

So without the assert i would be able to get a valid banner instead. What about removing the assert?

**Comment by knowyourcorp** [comment]
This isn't a production stable app. Lot of issues.

On Tue, Nov 13, 2012 at 1:17 PM, Riccardo Magliocchetti <
notifications@github.com> wrote:

> Just got ~50 assertion in production, no fun. These are the actual banners
> on my site, l is the list of all the banner weight of the same place.
> 
>  l = [10, 10, 5, 3, 5, 10, 9]
> s = sum(l)
> s
> 52
> bias = [x/float(s) for x in l]
> bias
> [0.19230769230769232, 0.19230769230769232, 0.09615384615384616,
> 0.057692307692307696, 0.09615384615384616, 0.19230769230769232,
> 0.17307692307692307]
> sum(bias)
> 1.0
> 
> from random import random
> from decimal import Decimal
> number = Decimal("%.18f" % random())
> current = Decimal(0)
> i = 0
> for x in bias:
> ... current += Decimal(x)
> ... if number <= current:
> ... print l[i]
> ... i += 1
> 9
> 
>  So without the assert i would be able to get a valid banner instead.
> What about removing the assert?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/martinogden/django-banner-rotator/issues/9.

## 

Grant Deken
CMO, Co-Founder

Phone: 305-330-1730 Ext. 701
Email: Grant@knowyourbank.com
Web: www.knowyourbank.com
Twitter: @knowyourbank http://www.twitter.com/knowyourbank |
@grdekenhttp://www.twitter.com/grdeken

This message is intended only for the designated recipient(s). It may
contain confidential or proprietary information and may be subject to
confidentiality protections. If you are not a designated recipient, you may
not review, copy or distribute this message. If you receive this in error,
please notify the sender by reply e-mail and delete this message. Thank you.

**Comment by xrmx** [comment]
Il 13/11/2012 19:46, knowyourcorp ha scritto:

> This isn't a production stable app. Lot of issues.

Then, it would be nice to file bugs so they can be fixed :)

**Comment by martinogden** [comment]
I'd like to keep the assert as the sum should always be 1. Rather, I'll investigate why the assertion is failing and write some tests.

**Comment by xrmx** [comment]
So this is the bias list which has 0.9998 as bias sum instead of 1.0:

[Decimal('0.1754'), Decimal('0.1754'), Decimal('0.0877'), Decimal('0.0526'), Decimal('0.0877'), Decimal('0.1754'), Decimal('0.1579'), Decimal('0.0877')]

The weight of these banners is respectively 10, 10, 5, 3, 5, 10, 9, 5

So it looks legit because of truncation / rounding / whatever that the sum is not exactly 1.0

**Comment by martinogden** [comment]
I removed the code that was rounding the floats and causing the invalid `bias_list` being passed to the pick method. This should now be fixed and there is no need to remove the assert.

**Comment by martinogden** [comment]
FIxed in `d5e3a264403fa6c536aaa6833eb6fd5c54d7cc5e`.
